### ISSUE TITLE
Allow to partially unwrap session iterator

### ISSUE THREAD

**Issue by fyfyrchik** [nspcc-dev/neo-go#3272]
# TLDR
Prefetch items in `CreateCallAndUnwrapIteratorScript()` or a new funtion.

# Is your feature request related to a problem? Please describe.

In general, there are 3 problems with session iterators:
1. For any number of items we must perform at least 2 round trips (get iterator and traverse it).
2. Hitting session limits. This is a real problem: in my setup we have 100-node cluster, each of the node has it's own contract and UI needs to query each of the contracts. It can also happen in public networks if, for example, many NeoFS nodes are connected to a single RPC-node (maybe less important with siamese nodes).
3. Sessions solve an important problem (consistent retrieval of a large collection of items), but have problems themselves: long-standing reading tx (current implementation detail) can prevent boltDB from growing, so avoiding them when possible could reduce the amount of resources we consume (for both bookkeeping/live tx in neo-go and network in general).

# Describe the solution you'd like

To support working with RPC nodes without session support we have [CreateCallAndUnwrapIteratorScript()](https://github.com/nspcc-dev/neo-go/blob/de98b39a9502d64ab797a9191891955d7b8df21e/pkg/smartcontract/entry.go#L21).

It is possible to extend it (or add a new function) to support prefetching.
The new `unwrap.*` function is trivial as are changes to the script (see spoiler below):
<details>
 <summary>NeoVM code of the trailing part</summary>

```go
        // 1st possibility: jump when the maximum number of items was reached.
	// In this case the iterator should be left on top, because unwrap.SessionIterator looks at the top item.
	// The partially read array will be the second item.
	retainIteratorOffset := script.Len()
	emit.Int(script.BinWriter, 2)
	emit.Opcodes(script.BinWriter, opcode.XDROP) // Remove maxIteratorResultItems from the 2-nd cell of estack, so that only iterator and resulting array are left on estack.
	emit.Opcodes(script.BinWriter, opcode.SWAP)
	emit.Opcodes(script.BinWriter, opcode.RET)

	// 2nd possibility: jump when the iterator has no more items.
	// In this case the top item will be stackitem.Null.
	// End of the program: push the result on stack and return.
	loadResultOffset := script.Len()
	emit.Int(script.BinWriter, 2)
	emit.Opcodes(script.BinWriter, opcode.XDROP) // Remove maxIteratorResultItems from the 2-nd cell of estack, so that only iterator and resulting array are left on estack.
	emit.Opcodes(script.BinWriter, opcode.NIP,   // Drop iterator as the 1-st cell on the stack.
		opcode.PUSHNULL) // Push null on top
	if err := script.Err; err != nil {
		return nil, fmt.Errorf("emitting iterator unwrapper script: %w", err)
	}

	// Fill in JMPIFNOT instruction parameter.
	bytes := script.Bytes()
	bytes[jmpIfNotOffset+1] = uint8(loadResultOffset - jmpIfNotOffset) // +1 is for JMPIFNOT itself; offset is relative to JMPIFNOT position.
	// Fill in jmpIfMaxReachedOffset instruction parameter.
	bytes[jmpIfMaxReachedOffset+1] = uint8(retainIteratorOffset - jmpIfMaxReachedOffset) // +1 is for JMPIF itself; offset is relative to JMPIF position.
	return bytes, nil
```
</details>

The question is whether we want to have such machinery in the neo-go repo itself. IMO yes, the optimization is quite common and it can be done even with the current interface, we just need to take some care in https://github.com/nspcc-dev/neo-go/blob/de98b39a9502d64ab797a9191891955d7b8df21e/pkg/rpcclient/invoker/invoker.go#L210-L225 -- currently it probihits having both `Values` and `ID` being not-nil simultaneously. With prefetching this can be possible, so we need to first process `Values` field.

# Describe alternatives you've considered
1. Increasing session limits AKA the perfect solution to every problem (no). It is not scalable, though 1k limit possibly can support 99% of all reasonable scenarious.
2. Perform `CreateCallAndUnwrapIteratorScript` and fallback on using sessions if we hit the limit. OK, but has more overhead for large collections of items. A bit more optimal on small collections, but the difference is negligible (<10 opcodes).
3. Having that function as a helper in client (not neo-go) code. Nice, but IMO everyone can benefit.

**Comment by roman-khimov** [comment]
I'd add some `smartcontract.CreateCallAndPrefetchIteratorScript` for this (one or two stack items returned, an array (always) and iterator if there are any items left) and an additional `unwrap.ArrayAndSessionIterator`. Modifying `CreateCallAndUnwrapIteratorScript` is not an option, it's supposed to be used for servers that have sessions disabled and you can't detect if they are from a script, so while it'll work until you reach the limit and then suddenly fail completely (server will return "sessions disabled" error). For `CreateCallAndPrefetchIteratorScript` we can specify that proper sessions _must_ be enabled. The other thing is that I don't exactly like mixing `iterator.Values` and `iterator.ID`, we have these options specified as mutually exclusive for quite some time and breaking this assumption can be dangerous, therefore an additional `unwrap` thing.

**Comment by fyfyrchik** [comment]
> we have these options specified as mutually exclusive for quite some time and breaking this assumption can be dangerous

What kind of problems can we encounter here? To me it seems perfectly backwards-compatible change. There _will_ be a new unwrap function, it can just reuse the same `result.Iterator` structure. The structure we are trying to construct was previously impossible to construct using our client library.

**Comment by roman-khimov** [comment]
`result.Iterator` has a direct mapping from/to returned stack item. But in your case the real result on the stack is much different from it (has an _additional_ array).
